kaiax/gov/headergov/impl/rewind.go

```diff
@@ -0,0 +1,24 @@
+package impl
+
+import (
+	"github.com/kaiachain/kaia/blockchain/types"
+	"github.com/kaiachain/kaia/common"
+)
+
+func (h *headerGovModule) RewindTo(newBlock *types.Block) {
+	// Remove entries from h.cache that are larger than num
+	h.cache.RemoveVotesAfter(newBlock.NumberU64())
+	h.cache.RemoveGovAfter(newBlock.NumberU64())
+}
+
+func (h *headerGovModule) RewindDelete(hash common.Hash, num uint64) {
+	h.cache.RemoveVotesAfter(num)
+	h.cache.RemoveGovAfter(num)
+
+	// Update stored block numbers for votes and governance
+	var voteBlockNums StoredUint64Array = h.cache.VoteBlockNums()
+	WriteVoteDataBlockNums(h.ChainKv, &voteBlockNums)
+
+	var govBlockNums StoredUint64Array = h.cache.GovBlockNums()
+	WriteGovDataBlockNums(h.ChainKv, &govBlockNums)
+}
```

Comment:
Is concurrency control unnecessary for the two types of block numbers, governance and vote?

- Write: `AddGov`
- Read: `IdxCache`
- Delete: `Rewind`

I believe two different actions can be performed simultaneously. Is that correct?

Reply:
There can be a race condition between these operations, thus leading to an inconsistency between DB and cache for a short time:

```
DB/cache write:  PostInsertBlock (append), RewindDelete (delete).
DB read:         Init.
Cache read:      API, VerifyHeader, Getters (EffectiveParam*).
```

If we assume `RewindDelete` is done in a single thread, I think there won't be a significant error even if we allowed DB/Cache inconsistency. However, I'm not 100% sure about the side effect.

I tried adding a mutex like below, but I don't like below style. I think to implement a consistent write-through cache, `HeaderCache` should handle mutex and DB update. Maybe I'll have to rename or refactor `HeaderCache`.

```diff
diff --git a/kaiax/gov/headergov/impl/api.go b/kaiax/gov/headergov/impl/api.go
index 6e731cf3..b1c01027 100644
--- a/kaiax/gov/headergov/impl/api.go
+++ b/kaiax/gov/headergov/impl/api.go
@@ -76,6 +76,9 @@ func (api *headerGovAPI) Vote(name string, value any) (string, error) {
 }
 
 func (api *headerGovAPI) IdxCache() []uint64 {
+	api.h.mu.RLock()
+	defer api.h.mu.RUnlock()
+
 	return api.h.cache.GovBlockNums()
 }
 
@@ -142,6 +145,9 @@ func (api *headerGovAPI) getParams(num *rpc.BlockNumber) (gov.PartialParamSet, e
 }
 
 func (api *headerGovAPI) Status() StatusResponse {
+	api.h.mu.RLock()
+	defer api.h.mu.RUnlock()
+
 	return StatusResponse{
 		GroupedVotes: api.h.cache.GroupedVotes(),
 		Governances:  api.h.cache.Govs(),
diff --git a/kaiax/gov/headergov/impl/consensus.go b/kaiax/gov/headergov/impl/consensus.go
index 5101b395..bdb9e10b 100644
--- a/kaiax/gov/headergov/impl/consensus.go
+++ b/kaiax/gov/headergov/impl/consensus.go
@@ -137,7 +137,8 @@ func (h *headerGovModule) checkConsistency(blockNum uint64, vote headergov.VoteD
 	return nil
 }
 
-// blockNum must be greater than epoch.
+// getExpectedGovernance does not return nil.
+// The epoch of blockNum must be greater than 0, because prev epoch index is required.
 func (h *headerGovModule) getExpectedGovernance(blockNum uint64) headergov.GovData {
 	prevEpochIdx := calcEpochIdx(blockNum, h.epoch) - 1
 	prevEpochVotes := h.getVotesInEpoch(prevEpochIdx)
@@ -150,7 +151,11 @@ func (h *headerGovModule) getExpectedGovernance(blockNum uint64) headergov.GovDa
 	return headergov.NewGovData(govs)
 }
 
+// getVotesInEpoch does not return nil.
 func (h *headerGovModule) getVotesInEpoch(epochIdx uint64) map[uint64]headergov.VoteData {
+	h.mu.RLock()
+	defer h.mu.RUnlock()
+
 	votes := make(map[uint64]headergov.VoteData)
 	for blockNum, vote := range h.cache.GroupedVotes()[epochIdx] {
 		votes[blockNum] = vote
diff --git a/kaiax/gov/headergov/impl/execution.go b/kaiax/gov/headergov/impl/execution.go
index 9e9d85cd..d99793aa 100644
--- a/kaiax/gov/headergov/impl/execution.go
+++ b/kaiax/gov/headergov/impl/execution.go
@@ -40,6 +40,9 @@ func (h *headerGovModule) PostInsertBlock(b *types.Block) error {
 }
 
 func (h *headerGovModule) HandleVote(blockNum uint64, vote headergov.VoteData) error {
+	h.mu.Lock()
+	defer h.mu.Unlock()
+
 	h.cache.AddVote(calcEpochIdx(blockNum, h.epoch), blockNum, vote)
 
 	var data StoredUint64Array = h.cache.VoteBlockNums()
@@ -59,17 +62,13 @@ func (h *headerGovModule) HandleVote(blockNum uint64, vote headergov.VoteData) e
 }
 
 func (h *headerGovModule) HandleGov(blockNum uint64, gov headergov.GovData) error {
-	h.cache.AddGov(blockNum, gov)
+	h.mu.Lock()
+	defer h.mu.Unlock()
 
-	// merge gov based on latest effective params.
-	gp := h.EffectiveParamSet(blockNum)
-	err := gp.SetFromMap(gov.Items())
-	if err != nil {
-		logger.Error("kaiax.HandleGov error setting paramset", "blockNum", blockNum, "gov", gov, "err", err, "gp", gp)
-		return err
-	}
+	h.cache.AddGov(blockNum, gov)
 
 	var data StoredUint64Array = h.cache.GovBlockNums()
 	WriteGovDataBlockNums(h.ChainKv, &data)
+
 	return nil
 }
diff --git a/kaiax/gov/headergov/impl/getter.go b/kaiax/gov/headergov/impl/getter.go
index 9f97c72c..3a1e8f90 100644
--- a/kaiax/gov/headergov/impl/getter.go
+++ b/kaiax/gov/headergov/impl/getter.go
@@ -17,6 +17,9 @@ func (h *headerGovModule) EffectiveParamSet(blockNum uint64) gov.ParamSet {
 }
 
 func (h *headerGovModule) EffectiveParamsPartial(blockNum uint64) gov.PartialParamSet {
+	h.mu.RLock()
+	defer h.mu.RUnlock()
+
 	ret := make(gov.PartialParamSet)
 	for _, num := range h.cache.GovBlockNums() {
 		if num > blockNum {
@@ -35,6 +38,9 @@ func (h *headerGovModule) NodeAddress() common.Address {
 }
 
 func (h *headerGovModule) GetGovernanceHistory() headergov.History {
+	h.mu.RLock()
+	defer h.mu.RUnlock()
+
 	return h.cache.History()
 }
 
diff --git a/kaiax/gov/headergov/impl/init.go b/kaiax/gov/headergov/impl/init.go
index b5083732..98f0436b 100644
--- a/kaiax/gov/headergov/impl/init.go
+++ b/kaiax/gov/headergov/impl/init.go
@@ -2,6 +2,7 @@ package impl
 
 import (
 	"math/big"
+	"sync"
 
 	"github.com/kaiachain/kaia/blockchain/state"
 	"github.com/kaiachain/kaia/blockchain/types"
@@ -41,10 +42,11 @@ type headerGovModule struct {
 
 	epoch uint64
 	cache *headergov.HeaderCache
+	mu    *sync.RWMutex // mutex for DB, cache consistency
 }
 
 func NewHeaderGovModule() *headerGovModule {
-	return &headerGovModule{}
+	return &headerGovModule{mu: &sync.RWMutex{}}
 }
 
 func (h *headerGovModule) Init(opts *InitOpts) error {
diff --git a/kaiax/gov/headergov/impl/rewind.go b/kaiax/gov/headergov/impl/rewind.go
index 8cdf99d4..248e2689 100644
--- a/kaiax/gov/headergov/impl/rewind.go
+++ b/kaiax/gov/headergov/impl/rewind.go
@@ -10,6 +10,9 @@ func (h *headerGovModule) RewindTo(newBlock *types.Block) {
 }
 
 func (h *headerGovModule) RewindDelete(hash common.Hash, num uint64) {
+	h.mu.Lock()
+	defer h.mu.Unlock()
+
 	h.cache.RemoveVotesAfter(num)
 	h.cache.RemoveGovAfter(num)
```

Reply:
On a second thought, I think no-mutex is ok:
- readers only use cache (except the initial DB read)
- writers update cache first, then DB

So DB write is a little delayed, but I think it won't cause any issues.

Reply:
I initally thought that the two regions (cache(memory) and DB) are the scope of context guard.

This would not significantly harm the operation and usage. However, we may have another complexity of thread-safe implementation later if we need to add or change something to the header module.

Could you calculate the consideration above and leave further comments?

Reply:
Do you mean a lock for each cache and DB respectively?

Reply:
Sure, a cache lock is for runtime consistency and DB lock is for a startup consistency. If you think it some different, please let me know.

Reply:
Added 56959249